vault/identity_store_util.go

```diff
@@ -0,0 +1,1244 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	// Insert or update entity in MemDB using the transaction created above
+	err = i.memDBUpsertEntityInTxn(txn, entity)
+	if err != nil {
+		return err
+	}
+
+	// Persist the entity object
+	err = i.storagePacker.PutItem(entity)
+	if err != nil {
+		return err
+	}
+
+	// Committing the transaction *after* successfully persisting entity
+	txn.Commit()
+
+	return nil
+}
+
+func (i *identityStore) loadEntities() error {
+	// Accumulate existing entities
+	i.logger.Debug("identity: loading entities")
+	existing, err := logical.CollectKeys(i.storagePacker.View())
+	if err != nil {
+		return fmt.Errorf("failed to scan for entities: %v", err)
+	}
+	i.logger.Debug("identity: entities collected", "num_existing", len(existing))
+
+	// Make the channels used for the worker pool
+	broker := make(chan string)
+	quit := make(chan bool)
+
+	// Buffer these channels to prevent deadlocks
+	errs := make(chan error, len(existing))
+	result := make(chan *storageBucketEntry, len(existing))
+
+	// Use a wait group
+	wg := &sync.WaitGroup{}
+
+	// Create 64 workers to distribute work to
+	for j := 0; j < consts.ExpirationRestoreWorkerCount; j++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case bucketKey, ok := <-broker:
+					// broker has been closed, we are done
+					if !ok {
+						return
+					}
+
+					bucketEntry, err := i.storagePacker.Get(bucketKey)
+					if err != nil {
+						errs <- err
+						continue
+					}
+
+					// Write results out to the result channel
+					result <- bucketEntry
+
+				// quit early
+				case <-quit:
+					return
+				}
+			}
+		}()
+	}
+
+	// Distribute the collected keys to the workers in a go routine
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for j, bucketKey := range existing {
+			if j%500 == 0 {
+				i.logger.Trace("identity: enities loading", "progress", j)
+			}
+
+			select {
+			case <-quit:
+				return
+
+			default:
+				broker <- bucketKey
+			}
+		}
+
+		// Close the broker, causing worker routines to exit
+		close(broker)
+	}()
+
+	// Restore each key by pulling from the result chan
+	for j := 0; j < len(existing); j++ {
+		select {
+		case err := <-errs:
+			// Close all go routines
+			close(quit)
+
+			return err
+
+		case bucketEntry := <-result:
+			// If there is no entry, nothing to restore
+			if bucketEntry == nil {
+				continue
+			}
+
+			for _, entity := range bucketEntry.Items {
+				fmt.Printf("loading entity: %#v\n", entity)
+				// Only update MemDB and don't hit the storage again
+				err = i.upsertEntity(entity, nil, false)
+				if err != nil {
+					return fmt.Errorf("failed to update entity in MemDB: %v", err)
+				}
+			}
+		}
+	}
+
+	// Let all go routines finish
+	wg.Wait()
+
+	if i.logger.IsInfo() {
+		i.logger.Info("identity: entities restored")
+	}
+
+	return nil
+}
+
+func (i *identityStore) upsertEntity(entity *entityStorageEntry, previousEntity *entityStorageEntry, persist bool) error {
```

Comment:
What's pretty unclear is the difference between upsertEntity and loadEntity as both cause upserts in memDB. Can you add some commenting explaining what each does and what the difference is?

As a corollary -- should they be different?